### PR TITLE
CMR-4489

### PR DESCRIPTION
**Pull request by ygliuvt**
As a Client User, I want to search services in CMR reference format.

### CODE REVIEW

File: index-set-app/src/cmr/index_set/services/index_service.clj [hunk 1]
```diff
@@ -1,23 +1,22 @@
...
+   [cmr.elastic-utils.connect :as es-util]
+   [cmr.index-set.config.elasticsearch-config :as es-config]
+   [cmr.index-set.data.elasticsearch :as es]
+   [cmr.index-set.services.messages :as m]
+   [cmr.transmit.echo.rest :as echo-rest])
+  (:import
+   (clojure.lang ExceptionInfo)))
+
+(def searchable-concept-types
+  "Defines the concept types that are indexed in elasticsearch and thus searchable."
```
Comment (oubiwann) [inline index-set-app/src/cmr/index_set/services/index_service.clj]: Thanks for this attention to (docstring) detail :-)

File: index-set-app/src/cmr/index_set/services/index_service.clj [hunk 2]
```diff
@@ -26,13 +25,13 @@
 (defn gen-valid-index-name
   "Join parts, lowercase letters and change '-' to '_'."
   [prefix-id suffix]
-  (s/lower-case (s/replace (format "%s_%s" prefix-id suffix) #"-" "_")))
+  (string/lower-case (string/replace (format "%s_%s" prefix-id suffix) #"-" "_")))
```
Comment (oubiwann) [inline index-set-app/src/cmr/index_set/services/index_service.clj]: @sxu123's got an open PR that's using `safe-lowercase` for lots of things now ... maybe everything? Worth checking with her about ...

File: indexer-app/src/cmr/indexer/data/index_set.clj [hunk 3]
```diff
@@ -637,6 +646,25 @@
    ;; associated collections stored as EDN gzipped and base64 encoded for retrieving purpose
    :collections-gzip-b64 (m/not-indexed (m/stored m/string-field-mapping))})
 
+(defmapping service-mapping :service
+  "Defines the elasticsearch mapping for storing services."
+  {:_id  {:path "concept-id"}}
+  {:concept-id (-> m/string-field-mapping m/stored m/doc-values)
+   :revision-id (-> m/int-field-mapping m/stored m/doc-values)
+   ;; This is used explicitly for sorting. The values take up less space in the
```
Comment (chris-durbin) [inline indexer-app/src/cmr/indexer/data/index_set.clj]: This comment is irrelevant now. By using doc-values we are not using the fielddata cache. This concept-seq-id was used when we were trying to reduce the size of the fielddata cache any way we could.

We should not use it for services as it's just increasing the size of the index and adding complexity.

File: search-app/src/cmr/search/api/concepts_search.clj [hunk 4]
```diff
@@ -33,7 +33,7 @@
 (defn- concept-type-path-w-extension->concept-type
   "Parses the concept type and extension (\"granules.echo10\") into the concept type"
   [concept-type-w-extension]
-  (-> #"^([^s]+)s(?:\..+)?"
```
Comment (chris-durbin) [inline search-app/src/cmr/search/api/concepts_search.clj]: Any idea what could break by this change? It looks fine to me, but why would we specifically avoid strings with the letter 's' except as the last letter. Madness.

File: search-app/src/cmr/search/api/services.clj [hunk 5]
```diff
@@ -8,6 +8,8 @@
 (def service-api-routes
   (context "/services" []
 
+    ;; Search for services route is defined in routes.clj
```
Comment (chris-durbin) [inline search-app/src/cmr/search/api/services.clj]: Why?

File: system-int-test/test/cmr/system_int_test/search/service/service_search_test.clj [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+              (join-fixtures
+               [(ingest/reset-fixture {"provguid1" "PROV1" "provguid2" "PROV2"})
+                services/grant-all-service-fixture]))
+
+(deftest search-for-services-validation-test
+  (testing "Unrecognized parameters"
+    (is (= {:status 400, :errors ["Parameter [foo] was not recognized."]}
+           (search/find-refs :service {:foo "bar"}))))
+
+  (testing "Unsupported sort-key parameters"
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/service/service_search_test.clj]: Do we have any supported sort keys? If so we should add a test and add to the documentation.

File: index-set-app/src/cmr/index_set/services/index_service.clj (hunk 2 above)
Comment (ygliuvt) [reply index-set-app/src/cmr/index_set/services/index_service.clj]: We should not be calling safe-lowercase for things that don't allow nil.

File: indexer-app/src/cmr/indexer/data/concepts/service.clj [hunk 7]
```diff
@@ -1,10 +1,49 @@
...
 
+(defmethod es/parsed-concept->elastic-doc :service
+  [context concept parsed-concept]
+  (let [{:keys [concept-id revision-id deleted provider-id native-id user-id
+                revision-date format extra-fields]} concept
+        {:keys [service-name]} extra-fields
+        concept-seq-id (:sequence-number (concepts/parse-concept-id concept-id))]
+    (if deleted
+      ;; This is only called by re-indexing (bulk indexing)
+      ;; Regular deleted variables would have gone through the index-service/delete-concept path.
```
[the remark this answers is not in the document]
Comment (ygliuvt) [reply indexer-app/src/cmr/indexer/data/concepts/service.clj]: Done.

File: indexer-app/src/cmr/indexer/data/index_set.clj (hunk 3 above)
Comment (ygliuvt) [reply indexer-app/src/cmr/indexer/data/index_set.clj]: Removed `concept-seq-id` field from mapping.

File: search-app/src/cmr/search/api/concepts_search.clj (hunk 4 above)
Comment (ygliuvt) [reply search-app/src/cmr/search/api/concepts_search.clj]: This regex is what caused CMR-4586. It doesn't work with `services` which start with an "s".

File: search-app/src/cmr/search/api/services.clj (hunk 5 above)
Comment (ygliuvt) [reply search-app/src/cmr/search/api/services.clj]: search by concept types is handled generically (sort of) in routes.clj for all concept types.

File: system-int-test/test/cmr/system_int_test/search/service/service_search_test.clj (hunk 6 above)
Comment (ygliuvt) [reply system-int-test/test/cmr/system_int_test/search/service/service_search_test.clj]: We don't yet. It is my plan to file an issue and work on it next.

**Pull request merged**